### PR TITLE
Add edge heuristic to avoid full cycle detection in common cases.

### LINKED ISSUE #7: Use heuristic for quick accept without cycle check in `add_edge`
The type invariant that the current DAG is acyclic can be used to speed up incremental checks. A full algorithm is still _O(|V| + |E|)_ though.

Some easy heuristics are:
- is either `a` or `b` without prior edges?
- is there already an edge from `a` to `b` with the same directionality?

### PR DESCRIPTION
**Pull request by mitchmindtree**
@bluss this adds the basic heuristic discussed in #7 

Closes #7.

### CODE REVIEW

**Comment by bluss** [comment]
Is it better to check the heuristic before or after adding the edge?

File: src/lib.rs [hunk 1]
```diff
@@ -469,6 +480,22 @@ impl<N, E, Ix = DefIndex> Dag<N, E, Ix> where Ix: IndexType {
...
+/// If our parent *a* has no parents or our child *b* has no children, or if there was already an
+/// edge connecting *a* to *b*, we know that adding this edge has not caused the graph to cycle.
+fn must_check_for_cycle<N, E, Ix>(dag: &Dag<N, E, Ix>,
+                                  a: NodeIndex<Ix>,
+                                  b: NodeIndex<Ix>,
+                                  new_edge: EdgeIndex<Ix>) -> bool
+    where Ix: IndexType,
+{
+    dag.parents(a).next(dag).is_some() && dag.children(b).next(dag).is_some()
+    && !dag.children(a).any(dag, |_, e, _| e != new_edge)
```
Comment (bluss) [inline src/lib.rs]: I think you mean to check that there is an edge from a to b here, needs a condition that the child is b?

Comment (mitchmindtree) [inline src/lib.rs]: Oh damn, good spot!

**Comment by mitchmindtree** [comment]
@bluss I think it is slightly better to check before, as it allows us to not have to `!=` the new index when checking for pre-existing edges.

**Pull request merged**